### PR TITLE
fix(browser): allow save source as raw content

### LINKED ISSUE #120: Improve file saving
Actually we have `saveSource` method to save source of the requested URL. However, I'm trying to save a zip file with this method, and it results to a corrupted zip (because of the prepend text content to the source file, https://github.com/zenstruck/browser/blob/1.x/src/Browser/Session.php#L98-L112)

It could be useful to have a dedicated method to save files as raw content (HTML or not).

### PR DESCRIPTION
**Pull request by welcoMattic**
Fix: https://github.com/zenstruck/browser/issues/120

We can now save source as raw, and make assertions on file like zip or any other binary format.

NOTE: we could improve assertions on the zip file itself by requiring `ext-zip` and use `\ZipArchive` class to open zip.

### CODE REVIEW

File: src/Browser/KernelBrowser.php [hunk 1]
```diff
@@ -466,6 +466,14 @@ final public function assertHtml(): self
         return $this->assertHeaderContains('Content-Type', 'html');
     }
 
+    /**
+     * @return static
+     */
+    final public function assertZip(): self
```
Comment (kbond) [inline src/Browser/KernelBrowser.php]: I don't think we should have this as part of the native API - I'm afraid this could open a flood of requests for a ton of them. IMO, this should be part of a custom browser.

What about `assertContentType()` instead?

**Review by kbond — CHANGES_REQUESTED** [review]
Sorry, I misunderstood the issue (didn't realize you actually wanted to save a raw zip file).

As an alternative to the `$raw` parameter, what do you think about only adding the meta-information if content type is `text/*`?

This way the saved file would always be valid to open in your OS's explorer. docx/xlsx/whatever

This would no longer add this meta-information to json so I'm not sure about this one... (@nikophil, wdyt?)

**Comment by welcoMattic** [comment]
@kbond I agree, we can base the raw process on the content type. For json, as the content type is `application/json` the metadata should be added, I'll write a test for it

**Comment by kbond** [comment]
Sounds good! For the `$raw` parameter, you can either remove it entirely (can always add later) or make it `?bool $raw = null`. When `null`, we make it true if content-type is `text/*` or `application/json`.

**Comment by nikophil** [comment]
not sure about the parameter if we implement the suggestion based on content type: if we have something like `application/zip` we'd **never** want the metadata to be added

**Comment by welcoMattic** [comment]
I've addressed all your comments @kbond @nikophil :wink:

File: src/Browser/Session.php [hunk 2]
```diff
@@ -95,26 +96,38 @@ public function expectException($expectedException, ?string $expectedMessage = n
...
-        $ret = "<!--\n";
+        $ret = '';
 
         try {
-            $ret .= "URL: {$this->getCurrentUrl()} ({$this->getStatusCode()})\n\n";
+            $isTextContent = str_contains((string) $this->getResponseHeader('content-type'), 'text')
+                || str_contains((string) $this->getResponseHeader('content-type'), 'json');
+        } catch (UnsupportedDriverActionException) {
+            $isTextContent = true;
+        }
```
Comment (nikophil) [inline src/Browser/Session.php]: please, could you comment out why this exception leads to `$isTextContent = true;`?

**Review by nikophil — CHANGES_REQUESTED** [review]
good for me, except the small comment I suggested

File: src/Browser/Session.php (hunk 2 above)
Comment (welcoMattic) [reply src/Browser/Session.php]: The exception is thrown by `Behat\Mink\Exception\UnsupportedDriverActionException: Response headers are not available from Zenstruck\Browser\Session\Driver\PantherDriver`, so to avoid BC break, I prefer set it to true, as it is actually

Comment (nikophil) [reply src/Browser/Session.php]: what happen if we use Panther and try to get a zip, then?

Comment (welcoMattic) [reply src/Browser/Session.php]: The zip file will be corrupted. 
The only way to avoid it using Panther is to use an optional `$raw` parameter

Comment (nikophil) [reply src/Browser/Session.php]: just to be sure, I'm not very used to Panther: when using Panther, we don't have any way to grab the headers of the response?

Comment (welcoMattic) [reply src/Browser/Session.php]: PantherDriver indirectly extends https://github.com/minkphp/Mink/blob/master/src/Driver/CoreDriver.php#L287 (through out https://github.com/welcoMattic/browser/blob/2cd69e9be35a50617a7a8ac9620c2f6e5d5eb35f/src/Browser/Session/Driver.php#L24), so I guess we don't have any way to retrieve response header from Panther? right @kbond ?

Moreover, Panther Client can't provider the Response object: https://github.com/symfony/panther/blob/main/src/Client.php#L144

**Comment by welcoMattic** [comment]
I rewrite the code to put the metadata prepended to the content under a condition via a global flag named `BROWSER_SOURCE_DEBUG` (set to false by default).

The only exception to this, is the call of `source()` method from `dump()` method, here we force the metadata to be prepended.

File: src/Browser/Session.php (hunk 2 above)
Comment (kbond) [reply src/Browser/Session.php]: Correct, we can't get the status code/headers from Panther.

File: src/Browser/KernelBrowser.php [hunk 3]
```diff
@@ -442,28 +442,36 @@ final public function assertHeaderContains(string $header, string $expected): se
         return $this;
     }
 
+    /**
+     * @return static
+     */
+    final public function assertContentType(string $contentType): self
```
Comment (kbond) [inline src/Browser/KernelBrowser.php:448]: Can you add this feature to the docs?

**Review by kbond — COMMENTED** [review]
I'm thinking we should forgo the global env variable for now.

Correct me if I'm wrong but when using the PHPUnit extension, it will no longer output the meta info in the generated files. I believe this is useful to help debug what path/headers caused the problem.

I don't believe there is any harm in always having the metadata in `text/*` files. My only question was for .json files as this doesn't output valid json... I am thinking this is ok as I believe most people would open .json files in a text editor/ide.

To summarize, I think we should always add metadata for text/* and application/json content types. All others should be the raw values.

**Comment by welcoMattic** [comment]
I'm actually writing tests, and I expect to be able to compare files to saved sources from Browser (zip, html, jpg, json). As I need to compare saved sources to static files I have in my tests folder, I would like to disable completely the metadata prepended to the sources.

But, in another hand, I would like to enable those if I need to debug something.

@nikophil WDYT about this general flag?

**Comment by kbond** [comment]
> I would like to disable completely the metadata prepended to the sources.

Got it, didn't realize you had a real use-case.

I'm sort of still thinking a `?bool $raw = null` parameter opposed to a global config. Would that work for your use-case?

**Comment by kbond** [comment]
> not sure about the parameter if we implement the suggestion based on content type: if we have something like application/zip we'd never want the metadata to be added

Agreed. What I am suggesting is to, by default, only add metadata to html,xml,plain-text/json content types.

**Comment by welcoMattic** [comment]
To make a decision, I think we must retrieve the original reason of the metadata presence? Was there a specific reason?

**Comment by kbond** [comment]
It's purpose is to make debugging x/html/json errors easier when:
1. Running `$browser->dump()/dd()` (shows the metadata in the console)
2. When using the PHPUnit extension to output files after a failure/error. You can view these files in your console/explorer or make available to view in your CI (github actions)

(2) above is I suppose the only hard reason to actually write the files with the metadata included. If you are calling `->saveSource()` yourself, you likely do not want the metadata.

**Comment by welcoMattic** [comment]
As it's for failure/error cases, even in CI, I think the `BROWSER_SOURCE_DEBUG` env var is quite appropriate, no?

Set to `false` most of the time, and toggle it to true in case of failure/error, if you need to debug some saved source files?

**Comment by kbond** [comment]
Yes, that's a fair point. I still think any non-text/json content should never have metadata added as it corrupts the file.

But I'm fine rolling with this _as is_ for now as, by default, this won't happen anymore.

File: src/Browser/Test/HasBrowser.php [hunk 4]
```diff
@@ -95,6 +96,7 @@ protected function browser(array $options = [], array $server = []): KernelBrows
 
         $browserOptions = [
             'source_dir' => $_SERVER['BROWSER_SOURCE_DIR'] ?? './var/browser/source',
+            'source_debug' => $_SERVER['BROWSER_SOURCE_DEBUG'] ?? false,
```
Comment (kbond) [inline src/Browser/Test/HasBrowser.php:99]: Can you add this option to https://github.com/zenstruck/browser#configuration?

**Review by kbond — CHANGES_REQUESTED** [review]

**Comment by welcoMattic** [comment]
I see your point, so I can also add a condition (in addition to `$sourceDebug`) to make a check on content-type, to never add metadata except for a fixed whitelist of content-type (which could grow up through time), is it ok?

**Comment by kbond** [comment]
Sure, we'll have to be loose with the check I think: `str_contains('text/', $contentType) || str_contains('json', $contentType)`

**Comment by welcoMattic** [comment]
Finally, I've added 2 conditions to writing metadata : `sourceDebug` config flag (via env var) and checking response header to determine if is text content or not.

Beside, we decided to never prepend source content with metadata if PantherBrowser is used (as we can't check on response headers with Panther response).

**Comment by kbond** [comment]
Thanks @welcoMattic!

File: src/Browser/Test/HasBrowser.php (hunk 4 above)
Comment (kbond) [reply src/Browser/Test/HasBrowser.php:99]: Added in https://github.com/zenstruck/browser/commit/a99150a8af3adb9a4f7113c94dd56ae4441d63d9.
